feat(delegator): Delegator Scaffolding Without Template Method and Strategy Patterns

maniosgrivei opened the pull request:
Just for comparison between both solutions:

With Template Method and Strategy Patterns: https://github.com/eu-sovereign-cloud/ecp/pull/98

Without Template Method and Strategy Patterns: this PR

Commit:
feat(delegator): remove old plugin stuff

Commit:
feat(delegator): resource handler types

Commit:
feat(delegator): exploratory block storage model

Commit:
feat(delegator): exploratory block storage delegator handler

Commit:
feat(delegator): exploratory solution without template method and strategy patterns

ionos-landgraf-vin commented on foundation/gateway/pkg/model/status.go:
```diff
@@ -0,0 +1,60 @@
+	s.State = StateCreating
+}
+
+func (s *ResourceStatus) SetAvtive() {
```
```suggestion
func (s *ResourceStatus) SetActive() {
```

ionos-landgraf-vin commented on foundation/gateway/pkg/model/status.go:
```diff
@@ -0,0 +1,60 @@
+	StateError     ResourceState = "error"
+)
+
+type ResourceStatus struct {
```
```mermaid
flowchart TD
    Start(( )) -->|Resource Submitted| Pending
    Pending -->|Reconciler Picks Up| Creating
    Creating -->|Creation Complete| Active
    Creating -->|Creation Failed| Error
    Active -->|Spec Changed| Updating
    Active -->|Delete Requested| Deleting
    Updating -->|Update Complete| Active
    Updating -->|Update Failed| Error
    Error -->|Retry if new| Creating
    Error -->|Retry if exists| Updating
    Error -->|Delete Requested| Deleting
    Deleting -->|Finalizers Complete| End(( ))

    classDef success fill:#10b981,color:#fff,stroke:#10b981
    classDef error fill:#ef4444,color:#fff,stroke:#ef4444
    classDef progress fill:#3b82f6,color:#fff,stroke:#3b82f6
    classDef point fill:#333,stroke:#333

    class Active success
    class Error error
    class Pending,Creating,Updating,Deleting progress
    class Start,End point

    linkStyle 0,1,2,4,5,6,11 stroke:#10b981,stroke-width:2px
    linkStyle 3,7 stroke:#ef4444,stroke-width:2px
    linkStyle 8,9,10 stroke:#f59e0b,stroke-width:2px

```

I think this struct should enforce this

maniosgrivei replied:
This is only a example/exploratory model, and it will be removed.

The real model for storage is the one conceived by @cristiGuranIonos on this PR: https://github.com/eu-sovereign-cloud/ecp/pull/137

ionos-landgraf-vin reviewed (approved):
RLGM

Commit:
refactor: rebase and adapt to the real block storage business model

cristiGuranIonos commented on foundation/delegator/pkg/adapter/resource_handler/block-storage.go:
```diff
@@ -0,0 +1,157 @@
+	case wantDelete(resource):
+		setState = h.repo.Delete
+
+	case wantIncreaseSize(resource):
```
I think this can be just a generic update

Commit:
fix: issues pointed in code review

  - only updade is sufficient (remove setStateFunc type and usages)
  - handle `pending` state
  - handle a single retry for creating and increase size operations

ionos-landgraf-vin commented on foundation/delegator/pkg/adapter/resource_handler/block-storage.go:
```diff
@@ -0,0 +1,161 @@
+	// Set the status of the resource properly.
+	if _, err := h.repo.Update(ctx, resource); err != nil {
+		return err // TODO: better error handling.
+	}
```
I was thinking that this might be better, so setResourceState is doing the update explicitly
```suggestion
		return h.setResourceState(ctx, resource, regional.ResourceStateActive)

	default:
		log.Fatal("must never achieve that condition")
	}
```

maniosgrivei replied:
To make short the discussion I made this commit : https://github.com/eu-sovereign-cloud/ecp/pull/107/commits/09ae06afd4685784ed9f7515a7248220e531061e

In case we re-consider to keep the concerns isolated, we only need to drop that one.

ionos-landgraf-vin reviewed (commented):
Rest LGTM

Commit:
refactor: move plugin handler to a proper folder + retry adjustments

Commit:
refactor: setResourceState now also update the resource on repo
